Import stuff

In [21]:
import dask.dataframe as dd
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

Import the file once

In [40]:
df = pd.read_csv('training_set_VU_DM_2014.csv')

Compute unique search_id

In [41]:
print (df['srch_id'].value_counts())

244169    38
149527    38
223395    38
250956    38
205230    37
77827     37
289360    37
313784    37
18010     37
216579    37
329684    37
158400    37
240149    37
87970     37
103906    37
140485    37
55214     37
199641    37
312060    37
211227    37
20060     37
60367     37
69052     37
6611      37
42558     37
101748    36
249242    36
224719    36
190202    36
202437    36
          ..
115328     5
283433     5
189439     5
290773     5
282577     5
78725      5
96978      5
264831     5
201665     5
191952     5
309216     5
57155      5
242382     5
233257     5
86237      5
19191      5
151326     5
324410     5
38759      5
280572     5
86926      5
310554     5
149308     5
118638     5
314644     5
194346     5
204271     5
228835     5
149306     5
324148     5
Name: srch_id, Length: 199795, dtype: int64


Compute unique date_time

In [42]:
print (df['date_time'].value_counts())

2012-11-02 14:04:51    95
2012-12-26 10:59:30    87
2013-05-15 18:39:51    70
2012-11-23 07:28:55    69
2013-01-13 18:45:07    68
2012-12-26 14:26:28    68
2012-12-03 10:33:44    68
2013-04-01 11:14:25    68
2013-06-16 10:43:33    68
2013-02-10 21:36:16    67
2013-06-12 07:35:06    67
2012-12-02 08:57:42    67
2012-12-03 11:45:52    67
2013-04-21 21:42:27    67
2013-03-13 10:38:00    67
2012-11-01 07:48:32    67
2013-06-18 17:54:30    67
2013-03-26 10:23:44    67
2013-06-29 22:04:06    67
2013-02-08 18:40:12    67
2013-02-15 09:19:23    67
2013-06-20 10:13:56    66
2013-06-16 07:52:09    66
2013-04-22 17:53:38    66
2013-03-27 06:30:24    66
2013-04-10 18:47:23    66
2013-06-26 08:07:07    66
2012-11-04 19:53:45    66
2013-02-27 19:14:13    66
2013-06-18 10:31:31    66
                       ..
2013-06-17 10:30:50     5
2013-01-27 18:38:30     5
2013-04-25 20:10:11     5
2013-04-14 09:25:46     5
2013-06-06 18:37:10     5
2013-04-23 21:56:00     5
2013-06-03 12:34:56     5
2013-03-27 1

As we see, there is some difference between the total amount of search id and date_time, so they cant be fully connected. If we want to work with it we probably have to filter these differences

Unique site_id

In [43]:
print (df['site_id'].value_counts())

5     3086006
14     481664
15     310788
24     237838
32     177330
18     127397
16      90032
12      73891
19      42758
9       42171
22      40262
11      34656
7       32067
29      29475
31      24420
26      16182
4       13499
25      12304
23      11620
1       10943
34       9944
10       9360
13       8205
17       6182
2        5540
27       5086
28       4318
30       3895
21       3654
3        3487
6        2522
33        663
20        157
8          31
Name: site_id, dtype: int64


Most searched from just a couple of sites. Maybe to strengthen the analysis we can leave out the least used sites.

Now we look at the unique countries of visitors

In [44]:
print (df['visitor_location_country_id'].value_counts())

219    2892489
100     495476
55      290461
216     217385
220     174296
129     123681
31       92636
99       67762
59       62066
92       45213
158      44505
103      35286
132      32740
117      25973
50       19623
39       18430
215      16361
181      12549
98       12024
13       11935
32       11836
15       11501
205      10736
81        9934
4         9428
33        9109
56        9070
23        8632
14        8239
73        8031
        ...   
122         56
27          56
168         55
116         55
46          55
188         55
87          52
139         48
193         47
121         46
183         39
174         33
210         33
166         33
228         33
67          32
94          32
217         31
64          31
3           31
54          31
58          30
198         30
101         28
66          27
226         25
190         24
146         15
83          13
150          8
Name: visitor_location_country_id, Length: 210, dtype: int64


No conclusions yet

Now at the unique countries of properties

In [45]:
print (df['prop_country_id'].value_counts())

219    3030125
100     311283
55      186465
31      154287
99      135388
129      91858
215      79981
220      71115
59       69453
216      61096
98       55075
158      42271
117      30682
205      29473
132      28694
39       28054
92       26709
103      24348
15       23548
32       19667
81       19381
4        19142
202      18850
14       18827
109      17868
73       17169
41       16525
56       15677
80       14454
225      14351
        ...   
192         57
172         57
136         55
197         38
25          35
51          33
141         32
108         29
82          29
167         25
19          24
89          23
190         23
135         22
217         21
185         21
226         21
111         20
46          19
62          18
79          16
128         15
140         15
87          13
208         13
180         12
7           10
126          7
204          5
165          2
Name: prop_country_id, Length: 172, dtype: int64


If we look at the first three countries of properties and the first three countries of properties it seems that a lot of people will go on vacation in their own country.

Let's look how many people want to go to the same destination as their origin (or where the search_id comes from)

In [46]:
df['samecountryasdestination'] = np.where((df['prop_country_id'] == df['visitor_location_country_id']), 1, 0)
print (df['samecountryasdestination'].value_counts())

1    3164018
0    1794329
Name: samecountryasdestination, dtype: int64


More people want to look for property in the same country as they are in people from other countries.

Some other things to create an overview for now

In [47]:
print (df['prop_id'].value_counts())

104517    2357
124342    2355
68420     2285
40279     2275
134154    2257
137997    2247
59781     2245
60846     2237
77089     2230
49656     2230
59657     2190
116942    2189
35223     2189
38419     2177
60468     2175
21018     2170
24545     2169
14082     2152
77795     2150
37818     2139
70177     2128
46274     2111
122112    2107
94455     2105
78500     2077
90845     2058
117294    1994
114932    1986
131892    1986
125083    1949
          ... 
28432        1
27677        1
25487        1
97632        1
111542       1
30295        1
10108        1
84720        1
135004       1
54171        1
137019       1
48472        1
70563        1
78759        1
78203        1
95151        1
65142        1
132033       1
98669        1
48135        1
18578        1
41878        1
15070        1
30294        1
99250        1
23823        1
7966         1
3868         1
38474        1
34363        1
Name: prop_id, Length: 129113, dtype: int64


We are talking about 129113 properties. We see a lot of those only got one visit. Possibly they are not really attractive in the search results. For the rest in the top 25 or so the differences between views are not really big.

A Closer look to how long people want to stay

In [48]:
print (df['srch_length_of_stay'].value_counts())

1     2143980
2     1203153
3      717298
4      373594
5      185744
7      117171
6      102671
8       33584
9       19224
10      16769
14      13282
11       8593
12       5992
13       4797
15       2961
16       1540
21       1513
17       1165
18        863
19        784
28        778
20        754
25        401
24        363
22        362
27        309
26        289
23        239
34         65
33         32
30         27
44         22
41         10
57          7
29          6
31          5
Name: srch_length_of_stay, dtype: int64


Most people want just a single night. And how longer, the less people are searching for it.

Are people going for brand hotels or for no-brand?

In [49]:
print (df['prop_brand_bool'].value_counts())

1    3147060
0    1811287
Name: prop_brand_bool, dtype: int64


Branded hotels have a preference. Good variable to account for.

In [51]:
print (df.columns)

Index(['srch_id', 'date_time', 'site_id', 'visitor_location_country_id',
       'visitor_hist_starrating', 'visitor_hist_adr_usd', 'prop_country_id',
       'prop_id', 'prop_starrating', 'prop_review_score', 'prop_brand_bool',
       'prop_location_score1', 'prop_location_score2',
       'prop_log_historical_price', 'position', 'price_usd', 'promotion_flag',
       'srch_destination_id', 'srch_length_of_stay', 'srch_booking_window',
       'srch_adults_count', 'srch_children_count', 'srch_room_count',
       'srch_saturday_night_bool', 'srch_query_affinity_score',
       'orig_destination_distance', 'random_bool', 'comp1_rate', 'comp1_inv',
       'comp1_rate_percent_diff', 'comp2_rate', 'comp2_inv',
       'comp2_rate_percent_diff', 'comp3_rate', 'comp3_inv',
       'comp3_rate_percent_diff', 'comp4_rate', 'comp4_inv',
       'comp4_rate_percent_diff', 'comp5_rate', 'comp5_inv',
       'comp5_rate_percent_diff', 'comp6_rate', 'comp6_inv',
       'comp6_rate_percent_diff', 'comp7_rate'

Just a heads-up for logistic regression trained on position (of same training set)

In [53]:
dftest = df['position']
df.drop('position')

st_time = t.clock()
Logreg = LogisticRegression()
Logreg.fit(df, dftest)
y_pred = Logreg.predict(df_test)
acc_log = round(Logreg.score(df, dftest)*100, 2)
print (acc_log, "--- %s seconds ---" % (t.clock() - st_time))

ValueError: labels ['position'] not contained in axis

Ideas:

See for each propery how many stars it has
See for each property the mean stars it has from reviews
See if there is descrepancy between stars and between stars from reviews.
pd.cut prices
Look at property id and position and search which values gave them that position.
Make training model based on position.
Regression uitvoeren en dan met correlaties kijken welke variabelen echt geen effect hebben
Prijzen van locaties binnen (in tientallen eerst)










>>> pd.cut(np.array([.2, 1.4, 2.5, 6.2, 9.7, 2.1]),
...        3, labels=["good", "medium", "bad"])
... 
[good, good, good, medium, bad, good]
Categories (3, object): [good < medium < bad]